### PR TITLE
23 volunteer checkin api

### LINKED ISSUE #23: Create volunteer check in api endpoint/server function
### Description
Create an API endpoint and corresponding server function that will be used to check in a volunteer at an event

### Technical Details
- The API route should be `/api/events/[eventId]/volunteers/[volunteerId]/check-in` meaning the API endpoint file should be at `src/api/[eventId]/volunteers/[volunteerId]/check-in/route.ts`
- Use the newly created Zod type from #22 to verify that the body of the check in request is in the correct format.
- Create a server function that creates an `EventVolunteer` at `src/server/actions/EventVolunteer.ts` and use this in the API endpoint to check in the volunteer.
- Then, use the update volunteer server function from #10 to update the checked in volunteers `previousRole` and `previousOrganization` to the organization and role they just checked in with.

### Dependencies
- #22 
- #10

### LINKED ISSUE #83: Make EventVolunteer objects unique
## Description
Currently, it is possible to check in the same volunteer twice. This is unwanted behavior. Add a unique index to prevent duplicate check-ins

## Technical Details
- Create an index on the EventVolunteer schema (`/server/models/EventVolunteer.ts`) that makes it so attempts to create records with an already-existing event & volunteer Id combination fail (one volunteer should only be able to check in to an event once)
  - [This](https://stackoverflow.com/questions/16061744/mongoose-how-to-define-a-combination-of-fields-to-be-unique) may be a good reference
- Update the POST `events/[eventId]/volunteers/[volunteerId]/check-in` API to handle this fail case if necessary (is also acceptable to add a TODO for the error handling story #58 

## Dependencies
- #23

### PR DESCRIPTION
**Pull request by andlrutt**
# Description

POST `events/[eventId]/volutneers/[volunteerId]/check-in` endpoint. Also added a unique index to event+volunteer id combinations, so that the same volunteer can not check in for the same event more than once

## Relevant issue(s)

- Closes #23 
- Closes #83 

## Questions

In `src/server/models/EventVolunteer.ts` I made `EventVolunteer.role` a required field, as it is listed as such in the schema diagram. Is this correct?
Please verify that my logic for the unique index on event+volunteer Ids is correct.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have assigned reviewers to this PR

### CODE REVIEW

**Comment by djwoun** [comment]
Is the logic: that a unique index of event+volunteer ID is made where no duplicate entries of the same event+volunteer ID can exist?

**Comment by djwoun** [comment]
Regarding the required roles, I think it's appropriate according to the schema diagram. Just a curious question but does CM want to track all of their volunteer roles, it may be a lot of work.

File: src/app/api/events/[eventId]/volunteers/[volunteerId]/check-in/route.ts [hunk 1]
```diff
@@ -1,7 +1,80 @@
...
+      return NextResponse.json(
+        { message: 'Invalid Volunteer Id' },
+        { status: 400 }
+      );
+    }
+
+    // Check Event ID
+    const validationEvent = zObjectId.safeParse(params.eventId);
+    if (!validationEvent.success) {
+      return NextResponse.json(
```
Comment (djwoun) [inline src/app/api/events/[eventId]/volunteers/[volunteerId]/check-in/route.ts:33]: Are we using CMerror?

Comment (djwoun) [inline src/app/api/events/[eventId]/volunteers/[volunteerId]/check-in/route.ts:33]: Are we using CMerror?

Comment (andlrutt) [reply src/app/api/events/[eventId]/volunteers/[volunteerId]/check-in/route.ts:33]: That'll be part of Dillon's PR -- not too concerned about errors in this one

**Comment by andlrutt** [comment]
> Is the logic: that a unique index of event+volunteer ID is made where no duplicate entries of the same event+volunteer ID can exist? 

Yes. There can be multiple records with the same volunteer ID and multiple records with the same event ID, but no two records can have the same volunteer AND event ID combination.

**Comment by andlrutt** [comment]
> Regarding the required roles, I think it's appropriate according to the schema diagram. Just a curious question but does CM want to track all of their volunteer roles, it may be a lot of work.

Yes, they want to know who did what role at what event. We're hoping we can ease the burden of entering that for every volunteer by defaulting their role to their previous role on check-in

**Comment by djwoun** [comment]
Ah ok, that makes sense. 
I think the logic makes sense in this scenario.

I was trying to test it on postman, and I think I am making a mistake with how I set the params. Any thoughts?
https://gyazo.com/14a6f5ee703a60555b1af6b4f010a0e7

**Comment by andlrutt** [comment]
> Ah ok, that makes sense. 
> I think the logic makes sense in this scenario.
> 
> I was trying to test it on postman, and I think I am making a mistake with how I set the params. Any thoughts?
> https://gyazo.com/14a6f5ee703a60555b1af6b4f010a0e7

replace [eventId] and [volunteerId] with the actual IDs in the url. Then go to the "Body" tab and make a JSON object with a role and optional organization id

**Comment by djwoun** [comment]
Ah, I see that works. It looks good to me.

**Review by djwoun — APPROVED** [review]

**Pull request merged**